src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -0,0 +1,354 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.InteropServices;
+
+namespace System.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// Managed SNI proxy implementation. Contains many SNI entry points used by SqlClient.
+    /// </summary>
+    internal class SNIProxy
+    {
+        public static readonly SNIProxy Singleton = new SNIProxy();
+
+        private GCHandle _gcHandle;
+
+        /// <summary>
+        /// Create consumer info
+        /// </summary>
+        /// <param name="async">Enable asynchronous support</param>
+        /// <param name="defaultBufferSize">Default buffer size</param>
+        /// <returns>Consumer info</returns>
+        public SNIConsumerInfo CreateConsumerInfo(bool async, int defaultBufferSize)
+        {
+            return new SNIConsumerInfo();
+        }
+
+        /// <summary>
+        /// Terminate SNI
+        /// </summary>
+        public void Terminate()
+        {
+        }
+
+        /// <summary>
+        /// Check if GC handle is allocated
+        /// </summary>
+        /// <returns></returns>
+        public bool IsGcHandleAllocated()
+        {
+            return _gcHandle.IsAllocated;
+        }
+
+        /// <summary>
+        /// Free GC handle
+        /// </summary>
+        public void FreeGcHandle()
+        {
+            _gcHandle.Free();
+        }
+
+        /// <summary>
+        /// Enable MARS support on a connection
+        /// </summary>
+        /// <param name="handle">Connection handle</param>
+        /// <returns>SNI error code</returns>
+        public uint EnableMars(SNIHandle handle)
+        {
+            if (SNIMarsManager.Singleton.CreateMarsConnection(handle) == TdsEnums.SNI_SUCCESS_IO_PENDING)
```

Comment:
CreateMarsConnection is going to throw an exception if handle was previously added already.  Is that expected, or did you want it to return SNI_ERROR in that case?